### PR TITLE
Feature :  symbol table Implementation

### PR DESCRIPTION
**Pull request by TheSYNcoder**
# Heading
Symbol Table Implementation

## Description
This PR is for the implementation of the symbol table.

### Remaining Tasks

- [x] Implement all methods
- [x] Fix clang errors
- [x] Fix memory leaks
- [x] Integrate with lexer.

### CODE REVIEW

File: src/include/symbol_table/symbole_table.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+namespace jucc {
+
+struct Node {
+  /**
+   * Used to store the name of the identifier
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbole_table.h]: use C++ class

File: src/include/symbol_table/symbole_table.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ public:
+  /**
+   * Store the identfier mappings with respect to their presence
+   * in different nesting levels in the program
+   */
+  std::map<std::string, Linked_list> hash_table_;
+  /**
+   * A vector to store different duplicate symbols found in the input
+   */
+  std::vector<std::string> duplicate_symbols;
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbole_table.h]: We had a discussion of using wrapper types instead of such long `std::vector<std::string>` => `jucc::grammar::Entity`.

Check the google docs.
Please wait till #4 gets merged.

**Review by bisakhmondal — COMMENTED** [review]
Cool.
Also please add the actual logic🤡😂

**Comment by bisakhmondal** [comment]
Also do rebase when you have some time.

File: src/include/symbol_table/symbole_table.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+   * One of int or float
+   */
+  std::string data_type_;
+  /**
+   * Used to store the nesting level of scoping
+   * Deeper the nesting, higher the scope
+   *
+   *      ...
+   *      int a0_;
+   *      if ( condition ){
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: if  (  condition  )  {
Please add the space. Please 🥺🙏

File: src/include/symbol_table/symbole_table.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+   *                     ...
+   *              }
+   *          }
+   *      }
+   *      ...
+   *      a0_ has 0 level , a1_ has 1 level and so on ...
+   */
+  std::string nesting_level_;
+  /**
+   * Constructir for node class
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: Typo in Constructor

File: src/include/symbol_table/symbole_table.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
+#ifndef JUCC_SYMBOL_TABLE_SYMBOL_TABLE_H
+#define JUCC_SYMBOL_TABLE_SYMBOL_TABLE_H
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+namespace jucc {
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: Use namespace symbol_table

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/include/symbol_table/symbole_table.h (hunk 2 above)
Comment (noob77777) [reply src/include/symbol_table/symbole_table.h]: member case is duplicate_symbols with _ suffix.
clang-tidy moment

File: src/include/symbol_table/symbole_table.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+   * Deletes the first node of the linked list
+   */
+  void DeleteStartingNode();
+  /**
+   * Returns true if linked list is empty or vice-versa
+   */
+  bool isEmpty();
+};
+
+class Symbol_table {
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: SymbolTable
readability-identifier-naming

File: src/include/symbol_table/symbole_table.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  std::string nesting_level_;
+  /**
+   * Constructir for node class
+   */
+  Node(std::string it_, std::string dt_, std::string nt_) : identifier_(it_), data_type_(dt_), nesting_level_(nt_) {}
+};
+
+typedef struct Node Node;
+
+class Linked_list {
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: LinkedList
readability-identifier-naming

File: src/include/symbol_table/symbole_table.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+   *      a0_ has 0 level , a1_ has 1 level and so on ...
+   */
+  std::string nesting_level_;
+  /**
+   * Constructir for node class
+   */
+  Node(std::string it_, std::string dt_, std::string nt_) : identifier_(it_), data_type_(dt_), nesting_level_(nt_) {}
+};
+
+typedef struct Node Node;
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: use using instead of typedef
modernize-use-using

File: src/include/symbol_table/symbole_table.h [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  void AddNewNode(Node *node_);
+  /**
+   * Deletes the first node of the linked list
+   */
+  void DeleteStartingNode();
+  /**
+   * Returns true if linked list is empty or vice-versa
+   */
+  bool isEmpty();
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: IsEmpty()
readability-identifier-naming

File: src/include/symbol_table/symbole_table.h [hunk 10]
```diff
@@ -0,0 +1,101 @@
+#ifndef JUCC_SYMBOL_TABLE_SYMBOL_TABLE_H
+#define JUCC_SYMBOL_TABLE_SYMBOL_TABLE_H
+
+#include <iostream>
+#include <map>
```
Comment (noob77777) [inline src/include/symbol_table/symbole_table.h]: is iostream necessary?
use unordered_map

File: src/include/symbol_table/symbole_table.h (hunk 1 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Why a struct seems good, why is a class needed ?

File: src/include/symbol_table/symbole_table.h (hunk 2 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Didn't get this...

File: src/include/symbol_table/symbole_table.h (hunk 10 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Iostream is not necessary can use unordered_map 👍

File: src/include/symbol_table/symbole_table.h (hunk 5 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Why ? 
Everything was supposed to be within namespace jucc !

File: src/include/symbol_table/symbole_table.h (hunk 3 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Space is there before and after "condition" 👀

File: src/include/symbol_table/symbole_table.h (hunk 4 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Sorry, okay will fix.

File: src/include/symbol_table/symbole_table.h (hunk 8 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Ahh man

File: src/include/symbol_table/symbole_table.h (hunk 7 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Yeah

File: src/include/symbol_table/symbole_table.h (hunk 2 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: `jucc::grammar::Entity` is used for a different context in productions which has nothing to do with the implementation of the symbol table.

File: src/include/symbol_table/symbole_table.h (hunk 3 above)
Comment (noob77777) [reply src/include/symbol_table/symbole_table.h]: )     SPACE   {

File: src/include/symbol_table/symbole_table.h (hunk 5 above)
Comment (noob77777) [reply src/include/symbol_table/symbole_table.h]: jucc::symbol_table, double namespace

Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: yeah doing it!

File: src/include/symbol_table/symbole_table.h (hunk 3 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbole_table.h]: Gotcha!

File: src/include/symbol_table/symbole_table.h (hunk 1 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbole_table.h]: uniformity, dear :)

**Comment by TheSYNcoder** [comment]
@noob77777 @bisakhmondal There is still clang-tidy problem

```
../src/include/grammar/grammar.h:88:3: error: an exception may be thrown in function '~Parser' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
  ~Parser();
  ^
/Users/shuvayan/Desktop/local/test/post/JuCC/src/grammar/grammar.cpp:32:9: error: an exception may be thrown in function '~Parser' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
Parser::~Parser() {
```

Please fix this ASAP.

PS : This push is after rebasing with `main`.

**Comment by noob77777** [comment]
> @noob77777 @bisakhmondal There is still clang-tidy problem
> 
> ```
> ../src/include/grammar/grammar.h:88:3: error: an exception may be thrown in function '~Parser' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
>   ~Parser();
>   ^
> /Users/shuvayan/Desktop/local/test/post/JuCC/src/grammar/grammar.cpp:32:9: error: an exception may be thrown in function '~Parser' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
> Parser::~Parser() {
> ```
> 
> Please fix this ASAP.
> 
> PS : This push is after rebasing with `main`.

This is a known issue with Mac OS please ignore for now. Or use // NOLINTNEXTLINE locally.

**Comment by TheSYNcoder** [comment]
@noob77777  @bisakhmondal 
Please review and merge.

File: src/include/lexer/lexer.h [hunk 11]
```diff
@@ -76,6 +80,23 @@ class Lexer {
...
+  /**
+   * The current nesting level as parsed by the lexer in
+   * the input file.
+   */
+  int current_nesting_level_{0};
+  /**
+   * vector to store duplicate symbol errors.
+   */
+  std::vector<std::string> duplicate_symbol_errors_;
+  /**
```
Comment (bisakhmondal) [inline src/include/lexer/lexer.h:95]: Please maintain uniformity. jucc::grammar::Entity

File: src/symbol_table/symbol_table.cpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+Node *LinkedList::CreateNewNode(std::string it_, std::string dt_, int nt_) {
+  Node *node = new Node(std::move(it_), std::move(dt_), nt_);
+  return node;
+}
+
+void LinkedList::AddNewNode(std::string it_, std::string dt_, int nt_) {
+  Node *node = LinkedList::CreateNewNode(std::move(it_), std::move(dt_), nt_);
+  node->next_ = head_;
+  head_ = node;
+}
```
Comment (bisakhmondal) [inline src/symbol_table/symbol_table.cpp:16]: R-value references!! possibility of memory leaking.

Take input with && directly.

File: src/include/symbol_table/symbol_table.h [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+  std::vector<std::string> undeclared_symbols_;
+
+ public:
+  /**
+   * Checks if the current identifier is present in the same nesting level
+   * int the hash_table. If present reports a duplicate symbol error, that is,
+   * inserts into the duplicate_symbols vector.
+   */
+  void CheckAndAddEntry(Node *node_);
+  /**
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h]: a single newlinr\e

File: src/include/symbol_table/symbol_table.h [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+   */
+  void InsertIntoDuplicateSymbols(const std::string &identifier_);
+
+  Node *GetLinkedListById(const std::string &id_);
+
+  std::vector<std::string> GetDuplicateSymbols();
+
+  std::vector<std::string> GetUndeclaredSymbols();
+
+  /**
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h]: Comments please

File: src/lexer/lexer.cpp [hunk 15]
```diff
@@ -23,21 +23,42 @@ int Lexer::GetToken(std::ifstream &is) {
       identifier_string_ += last_char;
     }
     if (identifier_string_ == "int") {
+      current_datatype_ = "int";
```
Comment (noob77777) [inline src/lexer/lexer.cpp]: current_datatype_  = identifier_string_
for create const static strings.
Just a suggestion.

File: src/include/symbol_table/symbol_table.h [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  void DeleteStartingNode();
+  /**
+   * Returns true if linked list is empty or vice-versa
+   */
+  bool IsEmpty();
+  /**
+   * returns head_
+   */
+  Node *GetHead();
+};
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h:83]: newline after each methods

File: src/include/symbol_table/symbol_table.h [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+   */
+  Node *next_;
+  /**
+   * Constructor for node class
+   */
+  Node(std::string it_, std::string dt_, int nt_)
+      : identifier_(std::move(it_)), data_type_(std::move(dt_)), nesting_level_(nt_), next_(nullptr) {}
+};
+
+using Node = struct Node;
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h]: redundant!! in C++ it it implicit

File: src/include/symbol_table/symbol_table.h [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace jucc {
+
+namespace symbol_table {
+
+struct Node {
+  /**
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h:14]: please use class, please

File: test/lexer/lexer_test.cpp [hunk 19]
```diff
@@ -197,9 +197,114 @@ TEST(lexer, lexer4) {
...
-  // UNEXPECTED BEHAVIOR AFTER THIS
-  // TOKENS received -2 -100 -2 -1
+  token = lex.GetToken(is);
+  ASSERT_EQ(jucc::lexer::TOK_ERROR, token);
+  token = lex.GetToken(is);
+  ASSERT_EQ(jucc::lexer::TOK_SEMICOLON, token);
+  token = lex.GetToken(is);
+  ASSERT_EQ(jucc::lexer::TOK_CURLY_CLOSE, token);
+  token = lex.GetToken(is);
+  ASSERT_EQ(jucc::lexer::TOK_COMMENT, token);
```
Comment (noob77777) [inline test/lexer/lexer_test.cpp:209]: ASSERT_EQ last token is TOK_EOF

File: src/include/symbol_table/symbol_table.h [hunk 20]
```diff
@@ -0,0 +1,138 @@
+#ifndef JUCC_SYMBOL_TABLE_H
+#define JUCC_SYMBOL_TABLE_H
+
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h]: Update header

File: src/include/symbol_table/symbol_table.h [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+  Node *GetHead();
+};
+
+class SymbolTable {
+  /**
+   * Store the identfier mappings with respect to their presence
+   * in different nesting levels in the program
+   */
+  std::unordered_map<std::string, LinkedList> hash_table_;
+  /**
```
Comment (bisakhmondal) [inline src/include/symbol_table/symbol_table.h:91]: If possible use a custom typedef which semantically conveys the logic

File: src/include/lexer/lexer.h [hunk 22]
```diff
@@ -76,6 +80,23 @@ class Lexer {
    * during tokenization.
    */
   double floatval_;
```
Comment (noob77777) [inline src/include/lexer/lexer.h:84]: Is there a specific reason all variables are public?
Most are not being accessed from other classes.
Use getters where applicable.

File: src/symbol_table/symbol_table.cpp [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+      if (hash_table_[id_nodes.first].IsEmpty()) {
+        // delete the entry
+        delete_queue.push_back(id_nodes.first);
+      }
+    }
+  }
+  for (const auto &id : delete_queue) {
+    if (hash_table_.find(id) != hash_table_.end()) {
+      hash_table_.erase(hash_table_.find(id));
+    }
```
Comment (bisakhmondal) [inline src/symbol_table/symbol_table.cpp]: Double querying. Optimize it

File: src/include/lexer/lexer.h [hunk 24]
```diff
@@ -76,6 +80,23 @@ class Lexer {
...
+  /**
+   * vector to store duplicate symbol errors.
+   */
+  std::vector<std::string> duplicate_symbol_errors_;
+  /**
+   * vector to store undeclared symbol errors.
+   */
+  std::vector<std::string> undeclared_symbol_errors_;
+
+  std::string current_datatype_;
```
Comment (noob77777) [inline src/include/lexer/lexer.h:106]: Uncommented variables.

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: test/grammar/grammar_test.cpp [hunk 25]
```diff
@@ -86,7 +86,7 @@ TEST(grammar, parser4) {
   ASSERT_EQ("grammar parsing error: invalid token %", parser.GetError());
 }
 
-TEST(grammar, parserTest5) {
+TEST(grammar, parser5) {
```
Comment (bisakhmondal) [inline test/grammar/grammar_test.cpp:89]: good :+1:

**Review by bisakhmondal — CHANGES_REQUESTED** [review]
:+1:

**Comment by noob77777** [comment]
Also this pr shows file changed by #4. I guess rebasing with main and force pushing should solve this.

File: src/include/symbol_table/symbole_table.h (hunk 2 above)
Comment (noob77777) [reply src/include/symbol_table/symbole_table.h]: duplicate resolving this thread

File: src/lexer/lexer.cpp (hunk 15 above)
Comment (TheSYNcoder) [reply src/lexer/lexer.cpp]: What is to be defined static ?

File: src/include/lexer/lexer.h (hunk 22 above)
Comment (TheSYNcoder) [reply src/include/lexer/lexer.h:84]: It is used for in testing. 
Changing it would require huge changes, but ok will try

File: src/include/lexer/lexer.h (hunk 24 above)
Comment (TheSYNcoder) [reply src/include/lexer/lexer.h:106]: yeah

File: src/include/lexer/lexer.h (hunk 11 above)
Comment (TheSYNcoder) [reply src/include/lexer/lexer.h:95]: Okay let's see.

File: src/symbol_table/symbol_table.cpp (hunk 12 above)
Comment (TheSYNcoder) [reply src/symbol_table/symbol_table.cpp:16]: std::move solves that purpose. 👀

File: src/include/symbol_table/symbol_table.h (hunk 13 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h]: check-lint didn;t complain 👀

File: src/include/symbol_table/symbol_table.h (hunk 14 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h]: yes

File: src/include/symbol_table/symbol_table.h (hunk 17 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h]: ohh

File: src/include/symbol_table/symbol_table.h (hunk 18 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h:14]: No struct is better for this purpose. I insist it staying as a struct.

File: src/include/symbol_table/symbol_table.h (hunk 20 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h]: To what ?

File: src/symbol_table/symbol_table.cpp (hunk 23 above)
Comment (TheSYNcoder) [reply src/symbol_table/symbol_table.cpp]: Okay

File: src/symbol_table/symbol_table.cpp (hunk 12 above)
Comment (bisakhmondal) [reply src/symbol_table/symbol_table.cpp:16]: it solves nothing!! Creates a problem if not handled it_ and dt_ properly in the rest of the code. 
Just do make sure they are untouched after AddNewNode invocation.

File: src/lexer/lexer.cpp (hunk 15 above)
Comment (noob77777) [reply src/lexer/lexer.cpp]: resolve

File: src/include/symbol_table/symbol_table.h (hunk 13 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h]: bcz it didn't detect any linting errors :)
Newline pls

Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h]: increases readability

File: src/include/symbol_table/symbol_table.h (hunk 18 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h:14]: I'm just asking for uniformity throughout the codebase. Both serve the same purpose.
Rest it's up to you ;)

Comment (noob77777) [reply src/include/symbol_table/symbol_table.h:14]: 🍿

File: src/include/symbol_table/symbol_table.h (hunk 20 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h]: JUCC_SYMBOL_TABLE_SYMBOL_TABLE_H
Namespaces, filename H

File: src/symbol_table/symbol_table.cpp (hunk 12 above)
Comment (TheSYNcoder) [reply src/symbol_table/symbol_table.cpp:16]: The function ends on line 16. How can they be used afterwards 👀

File: src/include/symbol_table/symbol_table.h (hunk 13 above)
Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h]: Fine.

File: src/symbol_table/symbol_table.cpp (hunk 12 above)
Comment (bisakhmondal) [reply src/symbol_table/symbol_table.cpp:16]: I was talking abt the parent scope.
Btw, after rechecking I feel it should fine. No worries :+1:

File: src/include/symbol_table/symbol_table.h (hunk 18 above)
Comment (noob77777) [reply src/include/symbol_table/symbol_table.h:14]: please fight

File: src/include/lexer/lexer.h (hunk 22 above)
Comment (noob77777) [reply src/include/lexer/lexer.h:84]: used for testing is not a valid reason

**Comment by TheSYNcoder** [comment]
@noob77777 @bisakhmondal 
Incorporated most of the changes. Other changes will be added as and when required as the codebase improves further.

TLDR,
PLEASE MERGE THIS BRANCH FOR GOD'S SAKE.
THANKS.

File: src/include/lexer/lexer.h (hunk 22 above)
Comment (noob77777) [reply src/include/lexer/lexer.h:84]: make private

Comment (TheSYNcoder) [reply src/include/lexer/lexer.h:84]: Moved it out of public , should be default to private. 👀

File: test/lexer/lexer_test.cpp (hunk 19 above)
Comment (noob77777) [reply test/lexer/lexer_test.cpp:209]: @TheSYNcoder not resolved

Comment (TheSYNcoder) [reply test/lexer/lexer_test.cpp:209]: F

File: src/include/symbol_table/symbol_table.h (hunk 18 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h:14]: please join

Comment (TheSYNcoder) [reply src/include/symbol_table/symbol_table.h:14]: NO

File: src/include/symbol_table/symbol_table.h (hunk 16 above)
Comment (bisakhmondal) [reply src/include/symbol_table/symbol_table.h:83]: please

**Review by noob77777 — APPROVED** [review]

**Review by bisakhmondal — APPROVED** [review]
OKAY :+1:
